Update ERC-7682: Add section on `capabilities.auxiliaryFunds` parameter, `requiredAssets`

adonesky1 opened the pull request:
This PR extends `ERC-7682` to specify that when a wallet supports the `auxiliaryFunds` capability, applications using `wallet_sendCalls` must include a `assetsRequiredForExecution` parameter nest under `capabilties` -> `auxiliaryFunds`. This parameter explicitly lists the assets and amounts required for successful execution, as transaction call data alone is insufficient for wallets to infer these requirements.

jiexi commented on ERCS/erc-7682.md:
```diff
@@ -86,9 +146,25 @@ An alternative we considered is defining a way for apps to fetch available auxil
+- The structure of `assetsRequiredForExecution` allows for future extension, such as specifying deadlines, slippage, or preferred bridges. Future versions of this standard MAY introduce additional fields as needed.
+
+### Decimals Source of Truth
+- Applications and wallets SHOULD verify the `decimals` value for each asset against onchain data or trusted sources to avoid errors from incorrect values.
```
in this case, why have the dapp provide `decimals` at all? to make sure that the value they're trying to request via the `amount` field is what they intend?

adonesky1 replied:
I've decided to remove the decimals property but add a section explaining that the wallet must cross reference the asset's decimals value to ensure it is interpreting values correctly:

https://github.com/ethereum/ERCs/pull/1126/commits/7fb4cf1d19c2e92a8bfbd10e8f948a858a9bca68

>The wallet MUST cross-check the `amount` value in `assetsRequiredForExecution` against the `decimal()` value specified on the `asset`'s contract to ensure that the amount is interpreted correctly according to the asset's decimal precision.

adonesky1 commented on ERCS/erc-7682.md:
```diff
@@ -72,10 +72,70 @@ type AuxiliaryFundsCapability = {
+  assetsRequiredForExecution?: Array<{
+    asset: `0x${string}`; // Asset address (use 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE for native asset)
+    amount: string;        // Amount required, as a string representing the integer value in the asset's smallest unit
+    decimals: number;      // Number of decimals for the asset (to interpret amount correctly)
```
Yeah I suppose this may be the right way. I suppose if we were going to add values that could otherwise be queried on the contract we'd maybe want to add the name/ticker. If we want remove it we'll just want to make sure its clear that the amount must be interpreted against the decimals that must be queried for.

adonesky1 replied:
This is making me wonder if we need to specify the standard (ERC20) of the asset

adonesky1 commented on ERCS/erc-7682.md:
```diff
@@ -72,10 +72,70 @@ type AuxiliaryFundsCapability = {
+          "amount": "1000000000000000000", // 1 DAI
+          "decimals": 18
+        },
+        {
+          "asset": "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE", // Native asset for gas (e.g., ETH)
+          "amount": "10000000000000000", // 0.01 ETH
+          "decimals": 18
+        }
```
Yeah I'm curious others thoughts on whether we should encourage a separate property where the dapp can atleast suggest what it believes the gas requirements will be

matthewwalsh0 replied:
I would argue that's entirely out of scope here as we're talking about assets that cannot be determined by the wallet, where as the gas limit and gas fee in the case of `wallet_sendCalls` is calculated by the wallet (pending some future capability).

adonesky1 commented on ERCS/erc-7682.md:
```diff
@@ -72,10 +72,87 @@ type AuxiliaryFundsCapability = {
-### App Implementation
+### Extended Usage: `assetsRequiredForExecution` Parameter
+
+When a wallet indicates support for the `auxiliaryFunds` capability (i.e., `supported: true`), applications that use the [`wallet_sendCalls`](./eip-5792.md#wallet_sendcalls) method MUST include an additional parameter named `assetsRequiredForExecution` in the call. 
```
modified [here](https://github.com/ethereum/ERCs/pull/1126/commits/6966ea9fefb9fe83e7b816d34189b5a96ca98bdd):

```
When a wallet indicates support for the `auxiliaryFunds` capability (i.e., `supported: true`), applications that use the [`wallet_sendCalls`](./eip-5792.md#wallet_sendcalls) method MAY include an `assetsRequiredForExecution` parameter in the `capabilities.auxiliaryFunds` object to enable the wallet to leverage this capability.
```

how does that sound?

lukasrosario replied:
^ kind of another argument for a separate capability. as a wallet, just expressing support for `auxiliaryFunds` is not expressive enough for an app to know if its `assetsRequiredForExecution` suggestion will be respected. at the very least, i think we need more than just `auxiliaryFunds: {supported: true}` in response to `wallet_getCapabilities`

adonesky1 commented on ERCS/erc-7682.md:
```diff
@@ -72,10 +72,87 @@ type AuxiliaryFundsCapability = {
+  }>;
+  capabilities?: {
+    auxiliaryFunds?: {
+      assetsRequiredForExecution?: Array<{
```
Yes I suppose that makes some sense 😅

adonesky1 replied:
Done here: https://github.com/ethereum/ERCs/pull/1126/commits/7860c89c879250ed0ca9e00cfd797580cb943ab2

adonesky1 commented on ERCS/erc-7682.md:
````diff
@@ -72,10 +72,87 @@ type AuxiliaryFundsCapability = {
+```
+
+The wallet, upon receiving this request, can:
+- Bridge or swap the required DAI from Chain Y to Chain X
````
good call out: https://github.com/ethereum/ERCs/pull/1126/commits/b0575c742552589104d49859c3674771c865419e

adonesky1 commented on ERCS/erc-7682.md:
```diff
@@ -72,10 +72,87 @@ type AuxiliaryFundsCapability = {
+Apps MUST include the `assetsRequiredForExecution` parameter in the `capabilities.auxiliaryFunds` object whenever they rely on the `auxiliaryFunds` capability to fulfill a user action. This ensures the wallet has the necessary information to provision assets as needed for successful execution.
+
+### Decimals Source of Truth
+- Applications and wallets SHOULD verify the `decimals` value for each asset against onchain data or trusted sources to avoid errors from incorrect values.
```
I've decided to remove the decimals property but add a section explaining that the wallet must cross reference the asset's decimals value to ensure it is interpreting values correctly:

https://github.com/ethereum/ERCs/pull/1126/commits/7fb4cf1d19c2e92a8bfbd10e8f948a858a9bca68


>The wallet MUST cross-check the `amount` value in `assetsRequiredForExecution` against the `decimal()` value specified on the `asset`'s contract to ensure that the amount is interpreted correctly according to the asset's decimal precision.

adonesky1 commented on ERCS/erc-7682.md:
```diff
@@ -72,10 +72,87 @@ type AuxiliaryFundsCapability = {
+  capabilities?: {
+    auxiliaryFunds?: {
+      assetsRequiredForExecution?: Array<{
+        asset: `0x${string}`; // Asset address (use 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE for native asset)
```
I'm not sure if we should require checksummed, we don't do that for any other APIs to my knowledge, so could create inconsistencies

jiexi commented on ERCS/erc-7682.md:
```diff
@@ -72,10 +72,104 @@ type AuxiliaryFundsCapability = {
+|----------|-----------------------------------|-------------|
+| -32080   | Auxiliary funds provisioning failed | Wallet attempted to provision funds but failed (e.g., bridge/swap error). |
+| -32081   | Asset not supported             | The requested asset is not available through the wallet’s auxiliary fund system. |
+| -32082   | Auxiliary funds timeout         | The provisioning operation took too long or timed out. |
```
does the dapp need to differentiate between -32080 and -32082?

adonesky1 replied:
no probably not! Cleaned up here: https://github.com/ethereum/ERCs/pull/1126/commits/dd3a05e0af065da03d7ce5a2819dce45c53d429f

jiexi commented on ERCS/erc-7682.md:
```diff
@@ -72,10 +72,104 @@ type AuxiliaryFundsCapability = {
+The wallet MUST cross-check the `amount` value in `assetsRequiredForExecution` against the `decimal()` value specified on the `asset`'s contract to ensure that the amount is interpreted correctly according to the asset's decimal precision.
+
+### Gas Estimation and Native Asset Handling
+- Applications SHOULD include an estimate for the required native asset (e.g., ETH for gas) in `assetsRequiredForExecution` when possible.
```
is this necessary if we've already determined the wallet capable of estimating the gas fee correctly without dapp input?

adonesky1 replied:
oh whoops I meant to remove this!

adonesky1 replied:
https://github.com/ethereum/ERCs/pull/1126/commits/dd3a05e0af065da03d7ce5a2819dce45c53d429f

cryptotavares commented on ERCS/erc-7682.md:
```diff
@@ -72,10 +72,100 @@ type AuxiliaryFundsCapability = {
+
+| Code     | Message                           | Description |
+|----------|-----------------------------------|-------------|
+| -32080   | Auxiliary funds provisioning failed | Wallet attempted to provision funds but failed (e.g., bridge/swap error). |
+| -32081   | Asset not supported             | The requested asset is not available through the wallet’s auxiliary fund system. |
+| -32082   | Auxiliary funds capability not available | The wallet no longer supports auxiliary funds on the requested chain. |
+| -32083   | Invalid assetsRequiredForExecution | The structure of the assetsRequiredForExecution object is malformed. |
```
I am wondering if the error codes should follow the pattern used in the [5792 spec](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-5792.md#error-codes), instead of the json-rpc 2.0 spec

vandan replied:
Yeah, we had something like that in [a draft MIP](https://github.com/MetaMask/metamask-improvement-proposals/pull/63/files) that corresponded to these changes. Maybe 5800 range to ensure that the errors remain distinct from the 5792 wallet_sendCalls errors.

adonesky1 replied:
Happy to migrate them to that number range!

adonesky1 replied:
done here: https://github.com/ethereum/ERCs/pull/1126/commits/c52d9b23375dc5557c84af09c7cebcfa2035fe1c

bumblefudge commented on ERCS/erc-7682.md:
```diff
@@ -72,10 +72,100 @@ type AuxiliaryFundsCapability = {
+
+### Amount Validation
+
+The wallet MUST cross-check the `amount` value in `assetsRequiredForExecution` against the `decimal()` value specified on the `asset`'s contract to ensure that the amount is interpreted correctly according to the asset's decimal precision.
```
this is a hard requirement to test... or rather, it's easy to check that a wallet is "checking" but hard to check what the correct interpretation is? it might be easier just to make this a SHOULD unless you're going to get more specific about what interpretation means here.

adonesky1 replied:
> this is a hard requirement to test... or rather, it's easy to check that a wallet is "checking" but hard to check what the correct interpretation is? it might be easier just to make this a SHOULD unless you're going to get more specific about what interpretation means here.

Maybe the way I phrased it is confusing. What I'm trying to say here is that the value amount that is specified by the dapp can only be correctly interpreted by the wallet (for representing to the user) if the wallet reads the decimal value of the contract in question. I had previously included decimals as a property in the `assetsRequiredForExecution` object so that the wallet wouldn't even need to check the contract to know how to interpret the amount, but after [some back and forth on this](https://github.com/ethereum/ERCs/pull/1126#discussion_r2220367032), I decided [to remove it](https://github.com/ethereum/ERCs/pull/1126/commits/7fb4cf1d19c2e92a8bfbd10e8f948a858a9bca68) and add this clause. Does my explanation here make it clearer? If so how do you think I should modify this clause to make it clearer?

adonesky1 commented on ERCS/erc-7682.md:
```diff
@@ -86,9 +176,16 @@ An alternative we considered is defining a way for apps to fetch available auxil
 
+## Backwards Compatibility
+- Applications SHOULD only include the `assetsRequiredForExecution` parameter if the wallet advertises the `auxiliaryFunds` capability.
+- If a wallet does not recognize or support the `assetsRequiredForExecution` parameter, it SHOULD ignore the parameter or return a clear error indicating lack of support. Apps SHOULD be prepared to handle such cases gracefully, potentially falling back to legacy behavior.
```
Removed here: https://github.com/ethereum/ERCs/pull/1126/commits/7e431dc85530f5b505eb2632d3f32da5a539936e

matthewwalsh0 commented on ERCS/erc-7682.md:
```diff
@@ -99,15 +99,56 @@ type WalletSendCallsRequest = {
+            tokenId: `0x${string}`; // Token ID as a hex string
+          };
+        };
+        erc1155?: {
```
Just confirming that we've gone with this standard as key approach to make it explicit that `erc20` doesn't support `tokenId`?

The alternative could be an array of shared schema such as:

```
assetsRequiredForExecution?: {
    asset: `0x${string}`;
    amount: `0x${string}`;
    tokenId?: `0x${string}`;
}[]
```

This has the advantage of simplicity, and meaning we wouldn't necessarily have to define specific standards in this specification, meaning it is decoupled and the wallet could support future standards?

One downside is the risk of duplicate assets.

adonesky1 replied:
Yeah this was closer to the format I'd originally proposed. I agree its probably cleaner than building in certain specification for the 3 primary token standards standards we anticipate. The main issue I see with it is that we aren't fully avoiding the possible need for extensions in the future since the properties we're assuming here (`amount` and `tokenId`) are the superset of the required properties for these specific standards. Any new standards with additionally required properties would need to propose an extension

adonesky1 replied:
Reverted to your suggested format: https://github.com/ethereum/ERCs/pull/1126/commits/55e7241c3fa69911ccd91b4d8642f542869dd9e9

adonesky1 commented on ERCS/erc-7682.md:
```diff
@@ -99,15 +99,56 @@ type WalletSendCallsRequest = {
+        };
+        erc721?: {
+          [asset: `0x${string}`]: {
+            tokenId: `0x${string}`; // Token ID as a hex string
```
done here: https://github.com/ethereum/ERCs/pull/1126/commits/e7e4fda0c5bee9e291e908eb267d91024890c5aa

lukasrosario commented on ERCS/erc-7682.md:
```diff
@@ -72,10 +72,118 @@ type AuxiliaryFundsCapability = {
+  capabilities?: {
+    auxiliaryFunds?: {
+      assetsRequiredForExecution?: {
+        address: `0x${string}`; // Asset contract address (MUST be checksummed)
```
i dont think we should include this constraint -- this is not the case anywhere else in 5792

adonesky1 replied:
Removed here: https://github.com/ethereum/ERCs/pull/1126/commits/b562bad687c0eeb975668d537134858b76e95b2e

lukaisailovic commented on ERCS/erc-7682.md:
```diff
@@ -72,10 +72,118 @@ type AuxiliaryFundsCapability = {
+  ],
+  "capabilities": {
+    "auxiliaryFunds": {
+      "assetsRequiredForExecution": [
```
Why does the dapp need to provide this? Wallet can just simulate the request

adonesky1 replied:
Simulation alone is not sufficient for a wallet to reliably determine which assets are required. If the user lacks the necessary balances or approvals on the target chain, the simulated call will revert before any token transfers occur, so no asset flow data is produced. Even with successful simulation, a wallet must know which contracts and methods to monitor; without full ABI knowledge of every contract involved, it cannot consistently decode the intended transfers. Some advanced engines allow “fake balance injection,” but that requires knowing which tokens and amounts to inject in the first place — the exact information the wallet is missing. The assetsRequiredForExecution parameter solves this chicken‑and‑egg problem by letting the dapp explicitly declare the required assets, so the wallet can provision them (via bridging, swapping, or funding) before executing the call.

adonesky1 commented:
> @adonesky1 dont think i have a super strong stance one way or the other currently, but just curious if you considered this as a separate capability? i actually think i agree with the overall direction here, but my first reaction was that this deserves its own capability.


So ATM I see `assetsRequiredForExecution` as an essential extension of `auxiliaryFunds` rather than a standalone capability. auxiliaryFunds signals what the wallet can do, while assetsRequiredForExecution provides the context it needs from the dapp in order to actually do it. Without this, the capability becomes much more constrained in practice.

For example, imagine a user wants to make a DAI deposit into Aave. The wallet has auxiliary funds it could provision, but at the moment holds no DAI on that chain. The calldata doesn’t reliably reveal which token or how much is being deposited, and simulations may revert without an ABI or because of runtime conditions. Without the dapp explicitly specifying the required assets, the wallet can’t act — even though it technically has the auxiliaryFunds capability. How do you see wallets meaningfully leveraging auxiliaryFunds in cases like this without assetsRequiredForExecution?

adonesky1 commented on ERCS/erc-7682.md:
````diff
@@ -72,10 +72,118 @@ type AuxiliaryFundsCapability = {
 ```
 
-### App Implementation
+### Extended Usage: `assetsRequiredForExecution` Parameter
````
see my [longer comment](https://github.com/ethereum/ERCs/pull/1126#issuecomment-3148540794) below on this. I think that `required` is actually the right language for what I think is needed here

adonesky1 commented on ERCS/erc-7682.md:
```diff
@@ -72,10 +72,118 @@ type AuxiliaryFundsCapability = {
+
+#### Example Scenario: Cross-Chain Deposit to Aave
+
+Suppose a user wants to deposit DAI into Aave on Chain X, but their wallet address on Chain X has no DAI or native gas token. However, the wallet has access to funds on Chain Y (e.g., Ethereum mainnet). The app, upon detecting the `auxiliaryFunds` capability, can construct a `wallet_sendCalls` request as follows:
```
> not following the premise here - auxiliaryFunds was introduced in order to signal offchain funds. In this example the funds are not just onchain but also on the same address just on another network. So the app, which needs to handle additional logic here, can already detect whether the wallet has enough funds - assuming the wallet is "connected".

So I think `auxIliaryFunds` ought to be considered to broadly indicate that funds can be provisioned from other sources whether they be offchain, other addresses in the users wallet, or from other chains from this same address. Why in your opinion should the `auxiliaryFunds` capability be limited to offchain funds only?

Essentially we want to provide a capability where the dapp shouldn't care about where the funds are or where they come from and they are providing the bare minimum information required by the wallet to ensure that its able to execute the transaction by gathering the assets it needs for the transaction being proposed to succeed. Normally right now dapps look at your assets on the particular chain the transaction is going to take place and don't even enable the flows for users to initiate transactions which they don't believe can be completed given the users onchain holdings for that chain. We want to say: "we can get funds from other sources, don't worry about where they're coming from, just tell us what we need for this transaction to succeed and we will provision those funds so the transaction can be executed."

jiexi commented on ERCS/erc-7682.md:
```diff
@@ -72,10 +72,118 @@ type AuxiliaryFundsCapability = {
+      assetsRequiredForExecution?: {
+        address: `0x${string}`; // Asset contract address (MUST be checksummed)
+        amount: `0x${string}`; // Amount required, as a hex string representing the integer value in the asset's smallest unit
+        standard: 'erc20' | 'erc721' | 'erc1155'; // Token standard
```
i know `0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE` exists, but it doesn't feel very elegant to represent native eth this way. Thoughts on adding a `'native'` standard here as well?

lukasrosario replied:
wallets can get this from `call.value` -- i would actually add something here that explicitly states wallets should just deduce this themselves @adonesky1

adonesky1 replied:
good call out @lukasrosario, adding something this effect

jxom commented on ERCS/erc-7682.md:
```diff
@@ -72,10 +72,118 @@ type AuxiliaryFundsCapability = {
+  }>;
+  capabilities?: {
+    auxiliaryFunds?: {
+      assetsRequiredForExecution?: {
```
Why can't this be named something simple like `required`? These capabilities already imply that they are intended "for execution", so the `*ForExecution` is kind of redundant.

adonesky1 replied:
I'd be fine with that!

adonesky1 replied:
Done here: https://github.com/ethereum/ERCs/pull/1126/commits/13bf16bad7223c59ae7c3dedf45b593c315cb0cd

adonesky1 commented:
@lukasrosario 

>i have one unaddressed comment above re getCapabilities response:
as a wallet, just expressing support for auxiliaryFunds is not expressive enough for an app to know if its assetsRequiredForExecution suggestion will be respected. at the very least, i think we need more than just auxiliaryFunds: {supported: true} in response to wallet_getCapabilities.
maybe something like auxiliaryFunds: { supported: true, executionFundingSuggestionSupported: true }?

[replied above
](https://github.com/ethereum/ERCs/pull/1126#discussion_r2255182275), but to restate it:

I see what you’re getting at, but I think it’s important to clarify: auxiliaryFunds is inherently a best‑effort capability, not a guarantee. Even if a wallet reports support, there’s no way to ensure in advance that it can provision the assets specified — that depends on factors like liquidity, bridging availability, and user approval. That’s why the spec defines error codes for provisioning failures.

Apps can always use assetsRequiredForExecution to make their needs explicit, but they must be prepared for cases where the wallet can’t fulfill them. Do you think requiring wallets to declare more in wallet_getCapabilities would add meaningful reliability here, given that the outcome is always conditional?

Or perhaps I'm misunderstanding your feedback

> i still don't feel great about the name. i understand the example you laid out, but again required still feels like a strong word for what is really a suggestion by the application. it doesn't even have to be malicious, it's very possible that an application miscalculates what is really "required". could we change to something like suggestion or tip?

I get the concern about calling them “required,” since the wallet can’t actually verify if the assets a dApp lists are truly necessary. On the other hand, calling them just “suggestions” feels too soft, because in the "happy path" the transaction won’t succeed without them.

Maybe the middle ground is: the dApp marks these as assets it says are needed, and the wallet shows that list to the user so they can decide whether to provision them. That way:
- The dApp makes clear what it thinks is required.
- The wallet isn’t pretending it can verify that claim (or perhaps in cases where it can verify it does)
- The user gets the final say after seeing what’s being asked for.

lukasrosario commented on ERCS/erc-7682.md:
```diff
@@ -98,10 +100,11 @@ type WalletSendCallsRequest = {
   }>;
   capabilities?: {
     auxiliaryFunds?: {
+      optional?: boolean;
```
maybe redundant? this is part of the capability definition. will leave it up to you but just fyi
<img width="492" height="174" alt="Screenshot 2025-08-14 at 5 50 45 PM" src="https://github.com/user-attachments/assets/d64144bf-94cf-4b52-a8c9-3a83fe26de46" />
